File: src/routes/photos.js
const express = require('express');
const supabase = require('../database/supabase-client');
const PhotoProcessingService = require('../services/photo-processing-service');
const { v4: uuidv4 } = require('uuid');
const router = express.Router();

// Create photo session
router.post('/session', async (req, res) => {
  try {
    const session_id = uuidv4();
    const { user_id, background_id, filter_id } = req.body;
    
    const { data, error } = await supabase
      .from('photo_sessions')
      .insert({
        session_id,
        user_id: user_id || null,
        background_id: background_id || null,
        filter_id: filter_id || null,
        created_at: new Date().toISOString()
      })
      .select();
    
    if (error) throw error;
    
    res.json({ success: true, session_id, data: data[0] });
  } catch (error) {
    console.error('Error creating session:', error);
    res.status(500).json({ error: error.message });
  }
});

// Upload & process photo
router.post('/upload', async (req, res) => {
  try {
    const { session_id, image_base64, background_id, filter_id } = req.body;
    
    if (!session_id || !image_base64) {
      return res.status(400).json({ error: 'session_id dan image_base64 required' });
    }
    
    // For now, just store metadata
    // Face detection will be implemented in Phase 4
    const photo_id = uuidv4();
    const { data, error } = await supabase
      .from('photos')
      .insert({
        photo_id,
        session_id,
        original_image_path: `uploads/${photo_id}.jpg`,
        processed_image_path: `uploads/${photo_id}_processed.jpg`,
        background_id,
        filter_id,
        face_count: 0,
        created_at: new Date().toISOString()
      })
      .select();
    
    if (error) throw error;
    
    res.json({
      success: true,
      photo_id,
      session_id,
      faces_detected: 0,
      data: data[0]
    });
  } catch (error) {
    console.error('Error uploading photo:', error);
    res.status(500).json({ error: error.message });
  }
});

// Get photos by session
router.get('/session/:session_id', async (req, res) => {
  try {
    const { data, error } = await supabase
      .from('photos')
      .select('*')
      .eq('session_id', req.params.session_id)
      .order('created_at', { ascending: false });
    
    if (error) throw error;
    
    res.json({ success: true, count: data.length, data });
  } catch (error) {
    console.error('Error fetching photos:', error);
    res.status(500).json({ error: error.message });
  }
});

// Get specific photo
router.get('/:photo_id', async (req, res) => {
  try {
    const { data, error } = await supabase
      .from('photos')
      .select('*')
      .eq('photo_id', req.params.photo_id)
      .single();
    
    if (error) throw error;
    if (!data) return res.status(404).json({ error: 'Photo not found' });
    
    res.json({ success: true, data });
  } catch (error) {
    console.error('Error fetching photo:', error);
    res.status(500).json({ error: error.message });
  }
});

// POST /api/photos/process - Full processing pipeline
router.post('/process', async (req, res) => {
  try {
    const { 
      photo_id, 
      background_id, 
      filter_id, 
      watermark_text,
      remove_background,
      blur_faces,
      pixelate_faces,
      frame_type
    } = req.body;
    
    if (!photo_id) {
      return res.status(400).json({ error: 'photo_id required' });
    }
    
    // Get photo from database
    const { data: photoData, error: photoError } = await supabase
      .from('photos')
      .select('*')
      .eq('photo_id', photo_id)
      .single();
    
    if (photoError) throw photoError;
    if (!photoData) return res.status(404).json({ error: 'Photo not found' });
    
    // Get filter config if needed
    let filterConfig = null;
    if (filter_id) {
      const { data: filterData } = await supabase
        .from('ai_filters')
        .select('filter_config')
        .eq('id', filter_id)
        .single();
      filterConfig = filterData?.filter_config ? JSON.parse(filterData.filter_config) : null;
    }
    
    // Process photo using the service
    const result = await PhotoProcessingService.processPhoto(
      photoData.original_image_path,
      {
        backgroundId: background_id,
        filterId: filter_id,
        filterConfig: filterConfig,
        watermarkText: watermark_text || 'Photobooth AI',
        removeBackground: remove_background || false,
        faces: [],
        frameType: frame_type,
        blurFaces: blur_faces || false,
        pixelateFaces: pixelate_faces || false
      }
    );
    
    // Update photo in database
    const { error: updateError } = await supabase
      .from('photos')
      .update({
        processed_image_path: result.processedPath,
        background_id,
        filter_id,
        processed_at: new Date().toISOString()
      })
      .eq('photo_id', photo_id);
    
    if (updateError) throw updateError;
    
    res.json({
      success: true,
      photo_id,
      processed_path: result.processedPath,
      steps: result.steps,
      message: `Photo processed in ${result.steps} steps`
    });
  } catch (error) {
    console.error('Error processing photo:', error);
    res.status(500).json({ error: error.message });
  }
});

// POST /api/photos/batch-process - Process multiple photos
router.post('/batch-process', async (req, res) => {
  try {
    const { photo_ids, background_id, filter_id, watermark_text } = req.body;
    
    if (!photo_ids || !Array.isArray(photo_ids) || photo_ids.length === 0) {
      return res.status(400).json({ error: 'photo_ids array required' });
    }
    
    const results = [];
    
    for (const photoId of photo_ids) {
      try {
        const { data: photoData } = await supabase
          .from('photos')
          .select('*')
          .eq('photo_id', photoId)
          .single();
        
        if (!photoData) continue;
        
        let filterConfig = null;
        if (filter_id) {
          const { data: filterData } = await supabase
            .from('ai_filters')
            .select('filter_config')
            .eq('id', filter_id)
            .single();
          filterConfig = filterData?.filter_config ? JSON.parse(filterData.filter_config) : null;
        }
        
        const result = await PhotoProcessingService.processPhoto(
          photoData.original_image_path,
          {
            backgroundId: background_id,
            filterId: filter_id,
            filterConfig: filterConfig,
            watermarkText: watermark_text || 'Photobooth AI'
          }
        );
        
        results.push({
          photo_id: photoId,
          success: result.success,
          processed_path: result.processedPath
        });
      } catch (error) {
        results.push({
          photo_id: photoId,
          success: false,
          error: error.message
        });
      }
    }
    
    res.json({
      success: true,
      processed: results.filter(r => r.success).length,
      failed: results.filter(r => !r.success).length,
      results
    });
  } catch (error) {
    console.error('Error batch processing photos:', error);
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;

File: backend_api/routers/photos.py
"""
Photo upload & processing router
"""
import secrets
import string
import uuid
import time
import threading
from io import BytesIO
from fastapi import APIRouter, UploadFile, File, Form, HTTPException, BackgroundTasks
from fastapi.responses import Response
from database import get_supabase
from schemas import ProcessPhotoRequest
from config import settings

router = APIRouter(prefix="/photobooth", tags=["photos"])

CLEANUP_DELAY_SECONDS = 300  # 5 minutes


def _generate_download_code(length: int = 6) -> str:
    return "".join(secrets.choice(string.ascii_uppercase + string.digits) for _ in range(length))


@router.post("/upload")
async def upload_photo(
    session_id: str = Form(...),
    photo: UploadFile = File(...),
):
    """Upload a photo to a session."""
    db = get_supabase()

    # Verify session exists and is active
    session_result = db.table("photo_sessions").select("id, status").eq("id", session_id).execute()
    if not session_result.data:
        raise HTTPException(404, "Session not found")
    if session_result.data[0]["status"] != "active":
        raise HTTPException(400, "Session is not active")

    # Validate file size
    contents = await photo.read()
    max_bytes = settings.max_upload_size_mb * 1024 * 1024
    if len(contents) > max_bytes:
        raise HTTPException(400, f"File too large. Max {settings.max_upload_size_mb}MB")

    # Count existing photos in session
    count_result = db.table("photos").select("id", count="exact").eq("session_id", session_id).execute()
    photo_number = (count_result.count or 0) + 1

    # Upload to Supabase Storage
    file_ext = photo.filename.split(".")[-1] if photo.filename and "." in photo.filename else "png"
    storage_path = f"{session_id}/{uuid.uuid4().hex}.{file_ext}"

    db.storage.from_("photos").upload(
        storage_path,
        contents,
        {"content-type": photo.content_type or "image/png"},
    )

    # Get public URL
    original_url = db.storage.from_("photos").get_public_url(storage_path)

    # Insert photo record
    photo_row = {
        "session_id": session_id,
        "original_path": storage_path,
        "original_url": original_url,
        "photo_number": photo_number,
        "status": "uploaded",
    }
    result = db.table("photos").insert(photo_row).execute()
    if not result.data:
        raise HTTPException(500, "Failed to save photo record")

    photo_data = result.data[0]

    # Run face expression detection in background
    _detect_face_expression_bg(photo_data["id"], contents)

    # Generate preview with removed background (for customization page)
    preview_url = None
    try:
        preview_url = _generate_preview_nobg(session_id, contents)
    except Exception as e:
        print(f"Preview BG removal failed: {e}")

    return {
        "success": True,
        "message": "Photo uploaded",
        "data": {
            "photo_id": photo_data["id"],
            "session_id": session_id,
            "original_url": original_url,
            "preview_nobg_url": preview_url,
            "photo_number": photo_number,
            "status": "uploaded",
        },
    }


def _generate_preview_nobg(session_id: str, image_bytes: bytes) -> str:
    """Generate a background-removed preview and return as base64 data URL."""
    import base64
    from ml.AI_Enhancement import remove_background
    from PIL import Image

    img = Image.open(BytesIO(image_bytes))
    nobg = remove_background(img)

    buf = BytesIO()
    nobg.save(buf, format="PNG")
    b64 = base64.b64encode(buf.getvalue()).decode("ascii")
    return f"data:image/png;base64,{b64}"


def _detect_face_expression_bg(photo_id: str, image_bytes: bytes):
    """Run face expression detection (sync, called after upload)."""
    try:
        from ml.face_expression import detect_expression
        from PIL import Image
        from io import BytesIO

        img = Image.open(BytesIO(image_bytes))
        expressions = detect_expression(img)

        if expressions:
            db = get_supabase()
            best = expressions[0]

            # Save to face_expressions table
            db.table("face_expressions").insert({
                "photo_id": photo_id,
                "expression": best["expression"],
                "confidence": best["confidence"],
                "bbox_x1": best.get("bbox", [0, 0, 0, 0])[0],
                "bbox_y1": best.get("bbox", [0, 0, 0, 0])[1],
                "bbox_x2": best.get("bbox", [0, 0, 0, 0])[2],
                "bbox_y2": best.get("bbox", [0, 0, 0, 0])[3],
            }).execute()

            # Update photo with primary expression
            db.table("photos").update({
                "face_expression": best["expression"],
                "face_confidence": best["confidence"],
            }).eq("id", photo_id).execute()

    except Exception as e:
        print(f"Face expression detection failed: {e}")


@router.post("/process")
async def process_photo(body: ProcessPhotoRequest):
    """Process photo with background, mascot, and filter."""
    db = get_supabase()

    # Get photo
    photo_result = db.table("photos").select("*").eq("id", body.photo_id).execute()
    if not photo_result.data:
        raise HTTPException(404, "Photo not found")
    photo = photo_result.data[0]

    # Get background
    bg_result = db.table("backgrounds").select("*").eq("id", body.background_id).execute()
    if not bg_result.data:
        raise HTTPException(404, "Background not found")
    background = bg_result.data[0]

    # Get mascot
    mascot_result = db.table("mascots").select("*").eq("id", body.mascot_id).execute()
    if not mascot_result.data:
        raise HTTPException(404, "Mascot not found")
    mascot = mascot_result.data[0]

    # Get filter (optional)
    ai_filter = None
    if body.filter_id:
        filter_result = db.table("ai_filters").select("*").eq("id", body.filter_id).execute()
        if filter_result.data:
            ai_filter = filter_result.data[0]

    # Create processing record
    processing_row = {
        "photo_id": body.photo_id,
        "background_id": body.background_id,
        "mascot_id": body.mascot_id,
        "filter_id": body.filter_id,
        "status": "pending",
        "started_at": "now()",
    }
    proc_result = db.table("photo_processing").insert(processing_row).execute()
    if not proc_result.data:
        raise HTTPException(500, "Failed to create processing record")

    processing = proc_result.data[0]
    processing_id = processing["id"]

    # ─── Run ML Pipeline ───
    try:
        start_time = time.time()

        # Update status: bg_removal
        db.table("photo_processing").update({"status": "bg_removal"}).eq("id", processing_id).execute()

        from ml.AI_Enhancement import process_photobooth
        from PIL import Image
        from io import BytesIO
        import requests as http_requests

        # Download original photo
        photo_url = photo["original_url"]
        resp = http_requests.get(photo_url, timeout=30)
        person_img = Image.open(BytesIO(resp.content))

        # Download background
        bg_url = db.storage.from_("assets").get_public_url(background["image_path"])
        bg_resp = http_requests.get(bg_url, timeout=30)
        bg_img = Image.open(BytesIO(bg_resp.content))

        # Download mascot
        mascot_url = db.storage.from_("assets").get_public_url(mascot["image_path"])
        mascot_resp = http_requests.get(mascot_url, timeout=30)
        mascot_img = Image.open(BytesIO(mascot_resp.content))

        # Update status: compositing
        db.table("photo_processing").update({"status": "compositing"}).eq("id", processing_id).execute()

        # Process
        filter_config = ai_filter.get("config", {}) if ai_filter else {}
        result_img = process_photobooth(person_img, bg_img, mascot_img, filter_config)

        # Update status: ai_enhance
        db.table("photo_processing").update({"status": "ai_enhance"}).eq("id", processing_id).execute()

        # Save result to storage
        output_buffer = BytesIO()
        result_img.save(output_buffer, format="PNG", quality=98, optimize=True)
        output_bytes = output_buffer.getvalue()

        result_path = f"{photo['session_id']}/{uuid.uuid4().hex}_processed.png"
        db.storage.from_("processed").upload(
            result_path,
            output_bytes,
            {"content-type": "image/png"},
        )
        processed_url = db.storage.from_("processed").get_public_url(result_path)

        processing_time_ms = int((time.time() - start_time) * 1000)

        # Generate download code & QR
        download_code = _generate_download_code()
        download_url = processed_url

        from ml.qr_generator import generate_qr_code

        qr_bytes = generate_qr_code(download_url)
        qr_path = f"{photo['session_id']}/{download_code}_qr.png"
        db.storage.from_("qrcodes").upload(
            qr_path,
            qr_bytes,
            {"content-type": "image/png"},
        )
        qr_url = db.storage.from_("qrcodes").get_public_url(qr_path)

        # Update processing record
        db.table("photo_processing").update({
            "status": "completed",
            "processed_path": result_path,
            "processed_url": processed_url,
            "processing_time_ms": processing_time_ms,
            "completed_at": "now()",
        }).eq("id", processing_id).execute()

        # Update photo status
        db.table("photos").update({"status": "processed"}).eq("id", body.photo_id).execute()

        # Create download link
        db.table("download_links").insert({
            "processing_id": processing_id,
            "download_code": download_code,
            "qr_code_path": qr_path,
            "qr_code_url": qr_url,
            "download_url": download_url,
        }).execute()

        # Schedule auto-cleanup after 5 minutes
        session_id = photo["session_id"]
        _schedule_cleanup(session_id, processing_id, body.photo_id, result_path, qr_path)

        return {
            "success": True,
            "message": "Photo processed successfully",
            "data": {
                "processing_id": processing_id,
                "photo_id": body.photo_id,
                "processed_url": processed_url,
                "download_code": download_code,
                "download_url": download_url,
                "qr_code_url": qr_url,
                "processing_time_ms": processing_time_ms,
            },
        }

    except Exception as e:
        # Mark as failed
        db.table("photo_processing").update({
            "status": "failed",
            "error_message": str(e)[:500],
        }).eq("id", processing_id).execute()

        db.table("photos").update({"status": "failed"}).eq("id", body.photo_id).execute()

        raise HTTPException(500, f"Processing failed: {str(e)}")


@router.get("/photo/{photo_id}")
async def get_photo(photo_id: str):
    """Get photo details with processing info."""
    db = get_supabase()

    photo_result = db.table("photos").select("*").eq("id", photo_id).execute()
    if not photo_result.data:
        raise HTTPException(404, "Photo not found")

    photo = photo_result.data[0]

    # Get processing records
    proc_result = (
        db.table("photo_processing")
        .select("*, download_links(*)")
        .eq("photo_id", photo_id)
        .execute()
    )

    # Get face expressions
    expr_result = db.table("face_expressions").select("*").eq("photo_id", photo_id).execute()

    return {
        "success": True,
        "message": "Photo retrieved",
        "data": {
            **photo,
            "processing": proc_result.data or [],
            "face_expressions": expr_result.data or [],
        },
    }


@router.get("/download/{download_code}")
async def download_by_code(download_code: str):
    """Get download URL by code."""
    db = get_supabase()

    result = (
        db.table("download_links")
        .select("*")
        .eq("download_code", download_code)
        .eq("is_active", True)
        .execute()
    )
    if not result.data:
        raise HTTPException(404, "Download link not found or expired")

    link = result.data[0]

    if link["download_count"] >= link["max_downloads"]:
        raise HTTPException(410, "Download limit reached")

    # Increment download count
    db.table("download_links").update({
        "download_count": link["download_count"] + 1,
    }).eq("id", link["id"]).execute()

    return {
        "success": True,
        "message": "Download link retrieved",
        "data": {
            "download_url": link["download_url"],
            "qr_code_url": link["qr_code_url"],
            "download_code": link["download_code"],
            "downloads_remaining": link["max_downloads"] - link["download_count"] - 1,
        },
    }


# ── Auto-Cleanup ──────────────────────────────────────────

def _schedule_cleanup(
    session_id: str,
    processing_id: str,
    photo_id: str,
    processed_path: str,
    qr_path: str,
):
    """Schedule deletion of all session data after CLEANUP_DELAY_SECONDS."""

    def _do_cleanup():
        time.sleep(CLEANUP_DELAY_SECONDS)
        try:
            db = get_supabase()

            # 1. Delete files from storage
            try:
                db.storage.from_("processed").remove([processed_path])
            except Exception:
                pass
            try:
                db.storage.from_("qrcodes").remove([qr_path])
            except Exception:
                pass

            # Delete original photos + preview
            photos_result = db.table("photos").select("original_path").eq("session_id", session_id).execute()
            for p in photos_result.data or []:
                try:
                    db.storage.from_("photos").remove([p["original_path"]])
                except Exception:
                    pass
            # Delete preview
            try:
                db.storage.from_("photos").remove([f"{session_id}/preview_nobg.png"])
            except Exception:
                pass

            # 2. Delete database records (cascading from session)
            db.table("download_links").delete().eq("processing_id", processing_id).execute()
            db.table("photo_processing").delete().eq("photo_id", photo_id).execute()
            db.table("face_expressions").delete().eq("photo_id", photo_id).execute()
            db.table("photos").delete().eq("session_id", session_id).execute()
            db.table("photo_sessions").update({"status": "expired"}).eq("id", session_id).execute()

            print(f"[Cleanup] Session {session_id} cleaned up after {CLEANUP_DELAY_SECONDS}s")
        except Exception as e:
            print(f"[Cleanup] Failed for session {session_id}: {e}")

    thread = threading.Thread(target=_do_cleanup, daemon=True)
    thread.start()
